packages/firestore/src/core/sync_engine.ts

```diff
@@ -343,6 +343,25 @@ export class SyncEngine implements RemoteSyncer {
     });
   }
 
+  /**
+   * Applies an OnlineState change to the sync engine and notifies any views of
+   * the change.
+   */
+  applyOnlineStateChange(onlineState: OnlineState) {
+    const newViewSnapshots = [] as ViewSnapshot[];
+    this.queryViewsByQuery.forEach((_, queryView) => {
```

Comment:
I feel like we should keep track of the online state somewhere (and this would be a good place). Then we could either assert that is has indeed changed or only raise events if it did.

Reply:
Can you elaborate?  I'm not sure what the benefit of tracking the onlineState here is just to assert that it changes.  This code doesn't actually rely on it having changed and I'd rather not add unnecessary state (as a general rule).

FWIW- We already track the current onlineState in RemoteStore, EventManager, and QueryListener, which is too many places in my opinion. :-)

Reply:
I just want to make sure that subsequent calls to disableNetwork() only trigger one snapshot event.

Reply:
I think we're pretty safe since we are guaranteed to only raise an OnlineState event if it actually changed. https://github.com/firebase/firebase-js-sdk/blob/master/packages/firestore/src/remote/remote_store.ts#L195

And the View code I added is also robust against duplicate OnlineState.Failed events since this.current will already be false here: https://github.com/firebase/firebase-js-sdk/pull/358/files#diff-59002bb6a4e41ae4ef48ff413744e9d3R277

Reply:
Good point. Guess we won't be adding a third layer of protection then. LGTM :)